libraries/chain/controller.cpp

```diff
@@ -4301,7 +4301,15 @@ struct controller_impl {
    }
 
    std::optional<finality_data_t> head_finality_data() const {
-      return apply_s<std::optional<finality_data_t>>(chain_head, [](const block_state_ptr& head) { return head->get_finality_data(); });
+      if (fork_db.version_in_use() == fork_database::in_use_t::both) {
+         // During transition to Savanna, need to fetch the head from fork_db which
+         // contains Savanna information for finality_data
+         return fork_db.apply_s<std::optional<finality_data_t>>([&](const auto& forkdb) {
+            return forkdb.head()->get_finality_data(); });
```

Comment:
We don't populate forkdb_s as we go because we only wanted the branch with IF Critical Block in it.

Reply:
Oh I thought we populate both DBs while in transition. So we have to store all information required by `base_digest` (including `header`, `core`, `finalizer_policies`, ...) in `block_header_state_legacy`?

Reply:
That or build it up right here from IF Genesis Block for each transition block.

Reply:
Probably it is cleaner to just build here, to avoid clutter `block_header_state_legacy` with too much less used data. The downside is to have to rebuild all previous Transition blocks for a particular block. But we should not have too many Transitions blocks. I can traverse the branch from IF Genesis block to current block in `forkdb_l`?

Reply:
Yes, similar to how it is done in replay.